### PR TITLE
Feature/sra v2 websocket server

### LINKED ISSUE #61: SRA v2: Websocket server
https://github.com/0xProject/standard-relayer-api/blob/master/ws/v2.md

### PR DESCRIPTION
**Pull request by kostromich**
resolve #61

### CODE REVIEW

File: app/src/helpers.js [hunk 1]
```diff
@@ -262,3 +262,16 @@ export const sendUnwrapWethTx = async (web3, account, wethToken, amount) => {
 
   return txHash
 }
+
+export const convertOrderToDexFormat = order => {
+  const decodedMakerAssetData = assetDataUtils.decodeERC20AssetData(order.makerAssetData)
+  const decodedTakerAssetData = assetDataUtils.decodeERC20AssetData(order.takerAssetData)
```
Comment (lunochkin) [inline app/src/helpers.js]: Use `decodeAssetDataOrThrow` method as it's more general.

File: app/src/modules/index.js [hunk 2]
```diff
@@ -151,46 +154,67 @@ export const setOrderbook = ({ bids: bidOrders, asks: askOrders }) => (dispatch,
...
 
-    dispatch(setAsks(asksSorted))
+  if (newAsks.length > 0) {
+    dispatch(setAsks(R.sort(R.descend(R.path(['data', 'price'])), [...asks, ...newAsks])))
   }
 
   dispatch(resetHighlighting())
 }
 
+// TODO get orderbook via http-request
```
Comment (lunochkin) [inline app/src/modules/index.js]: Implement orders loading with http at the same PR.

File: server/src/controllers/V0RelayerController.ts [hunk 3]
```diff
@@ -110,20 +111,20 @@ class V0RelayerController {
...
 
-    await this.orderRepository.save(order)
+    const decodedMakerAssetData = assetDataUtils.decodeAssetDataOrThrow(order.makerAssetData)
+    const decodedTakerAssetData = assetDataUtils.decodeAssetDataOrThrow(order.takerAssetData)
+    const orderToSave = {
+      ...order,
+      makerAssetProxyId: decodedMakerAssetData['assetProxyId'],
+      takerAssetProxyId: decodedTakerAssetData['assetProxyId'],
+      makerAssetAddress: decodedMakerAssetData['tokenAddress'],
+      takerAssetAddress: decodedTakerAssetData['tokenAddress']
```
[the remark this answers is not in the document]
Comment (lunochkin) [reply server/src/controllers/V0RelayerController.ts]: Also this code is the same as in frontend function `convertOrderToDexFormat`.
And it's better to move it to helpers as it was done in frontend.

File: server/src/wsServers/WsRelayerServer.ts [hunk 4]
```diff
@@ -17,43 +18,101 @@ class WsRelayerServer {
...
-            channel: 'orderbook',
-            requestId,
-            payload: {
-              asks,
-              bids
-            }
+        if (type === 'subscribe' && channel === 'orders') {
+          if (!requestId) {
+            log.info('requestId is missing')
+            ws.close(1011, 'requestId is missing')
```
[the remark this answers is not in the document]
Comment (kostromich) [reply server/src/wsServers/WsRelayerServer.ts]: Add HTTP response format for validation error

File: app/src/modules/index.js (hunk 2 above)
Comment (kostromich) [reply app/src/modules/index.js]: resolve #53

File: app/src/modules/index.js [hunk 5]
```diff
@@ -135,62 +139,109 @@ export const resetHighlighting = () => async (dispatch, getState) => {
...
 
-  if (order.makerAssetAddress === currentToken.address && order.takerAssetAddress === marketplaceToken.address) {
-    const bid = {
-      ...generateBid({ order, baseToken: marketplaceToken, quoteToken: currentToken }),
-      highlight: true
-    }
-    const bidsSorted = R.sort(R.descend(R.path(['data', 'price'])), [...bids, bid])
+  const expandedOrders = orders.map(order => ({
+    ...generateBid({
+      order: convertOrderToDexFormat(order),
```
Comment (lunochkin) [inline app/src/modules/index.js]: We don't need to use `convertOrderToDexFormat` here any more as we have it used inside `generateBid`

File: app/src/App.js [hunk 6]
```diff
@@ -18,20 +18,18 @@ class App extends React.Component {
...
 
-      if (type === 'snapshot' && channel === 'orderbook') {
-        store.dispatch(setOrderbook(payload))
-      }
-
-      if (type === 'update' && channel === 'orderbook') {
-        store.dispatch(addOrder(payload))
+      if (type === 'update' && channel === 'orders') {
+        // ignore metadata-key, pass list of sra2-orders
+        const orders = payload.map(one => one.order)
```
Comment (lunochkin) [inline app/src/App.js]: It seems to be inconsistent with `setOrderbook`.
The latter accepts `{ asks, bids }` object and each element of `asks` and `bids` has `order` property.
So it's better not to flatten payload here and pass it directly to `addOrders`.

**Pull request closed without merging**